README.md
# django_advanced_password_validation

[![test](https://github.com/ezrajrice/django_advanced_password_validation/actions/workflows/test.yaml/badge.svg?branch=main)](https://github.com/ezrajrice/django_advanced_password_validation/actions/workflows/test.yaml)
[![Coverage Status](https://coveralls.io/repos/github/ezrajrice/django_advanced_password_validation/badge.svg?branch=main)](https://coveralls.io/github/ezrajrice/django_advanced_password_validation?branch=main)

Extends Django password validation options to include minimum uppercase, minimum lowercase, minimum numerical, and minimum special characters. This was created in an attempt to keep up with industry standards for strong user passwords.

This package has been tested with python 3.9+.

## Prerequisites

Requires Django 3.2 or later.
You can install the latest version of Django via pip:

```bash
pip install django
```

Alternatively, you can install a specific version of Django via pip:

```bash
pip install django=3.2
```

> **_NOTE:_**  See the [django-project](https://docs.djangoproject.com) documentation for information on non-deprecated Django versions.

## Installation

### Normal installation

Install django_advanced_password_validation via pip:

```bash
pip install django_advanced_password_validation
```

### Development installation

```bash
git clone https://github.com/ezrajrice/django_advanced_password_validation.git
cd django_advanced_password_validation
pip install --editable .
```

### Usage

The optional validators must be configured in the settings.py file of your django project to be actively used in your project.

#### /my-cool-project/settings.py

```python
INSTALLED_APPS = [
    ...
    'django_advanced_password_validation',
    ...
]

AUTH_PASSWORD_VALIDATORS = [
    ...
    {
        'NAME': 'django_advanced_password_validation.advanced_password_validation.ContainsDigitsValidator',
        'OPTIONS': {
            'min_digits': 1
        }
    },
    {
        'NAME': 'django_advanced_password_validation.advanced_password_validation.ContainsUppercaseValidator',
        'OPTIONS': {
            'min_uppercase': 1
        }
    },
    {
        'NAME': 'django_advanced_password_validation.advanced_password_validation.ContainsLowercaseValidator',
        'OPTIONS': {
            'min_lowercase': 1
        }
    },
    {
        'NAME': 'django_advanced_password_validation.advanced_password_validation.ContainsSpecialCharactersValidator',
        'OPTIONS': {
            'min_characters': 1
        }
    },
    ...
]
```

### Options

Here is a list of the available options with their default values.

| Validator | Option | Default |
| --- |:---:| ---:|
| ContainsDigitsValidator | min_digits | 1 |
| ContainsUppercaseValidator | min_uppercase | 1 |
| ContainsLowercaseValidator | min_lowercase | 1 |
| ContainsSpecialCharactersValidator | min_characters | 1 |
| MaximumLengthValidator | max_length | 128 |
| MaxConsecutiveCharactersValidator | max_consecutive | 3 |
| ConsecutivelyIncreasingDigitValidator | max_consecutive | 3 |
| ConsecutivelyDecreasingDigitValidator | max_consecutive | 3 |

## Authors

* **Ezra Rice** - _Initial work_ - [ezrajrice](https://github.com/ezrajrice)

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details.

## Acknowledgments

* **Victor Semionov** - _Contributor_ - [vsemionov](https://github.com/vsemionov)
* **Mostafa Moradian** - _Contributor_ - [mostafa](https://github.com/mostafa)


HISTORY.md
# VERSION HISTORY

### VERSION 1.2.0 - 2023-11-17

**Added**

- N/A

**Removed**

- Django 2.2 no longer supported as it has been deprecated by the Django Project

**Edited**

- N/A

**Bug Fix**

- Fixed improper use of alias "_" for plural messages with ngettext
- Changed gettext usage to lazy translations

### VERSION 1.1.0 - 2022-05-23

**Added**

- MaximumLengthValidator
- MaxConsecutiveCharactersValidator
- ConsecutivelyIncreasingDigitValidator
- ConsecutivelyDecreasingDigitValidator

**Removed**

- N/A

**Edited**

- Updated the Options list to show inputs and default values for the new methods.

**Bug Fix**

- package has been renamed from "django-advanced_password_validation" to "django_advanced_password_validation" to fix the *django.core.exceptions.improperlyconfigured: the app label 'django-advanced_password_validation' is not a valid python identifier* error

### VERSION 1.0.4 - 2020-03-25

**Added**

- N/A

**Removed**

- Unused import gettext has been removed.

**Edited**

- *ContainsNumeralsValidator* has been modified to *ContainsDigitsValidator* to be a more intuitive naming convention (i.e. 15 is one numeral, but two digits)
  - Option *min_numerals* has been changed to *min_digits*

**Bug Fix**

- ContainsSpecialValidator was only checking for one (1) special character instead of the minimum parameter.

### VERSION 1.0.3 - 2020-03-20

**Added**

- ContainsNumeralsValidator
- ContainsUppercaseValidator
- ContainsLowercaseValidator
- ContainsSpecialCharactersValidator

django_advanced_password_validation/tests/settings.py
"""
Django settings for tests.
"""

import os
from typing import List

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SECRET_KEY = "SECRET"
DEBUG = True
ALLOWED_HOSTS: List[str] = []
INSTALLED_APPS = [
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "django_advanced_password_validation",
]
MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
]
ROOT_URLCONF = "urls"
TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [BASE_DIR + "/templates"],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]
WSGI_APPLICATION = "django_advanced_password_validation.wsgi.application"
DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.sqlite3",
        "NAME": os.path.join(BASE_DIR, "db.sqlite3"),
    }
}
AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
        "OPTIONS": {
            "min_length": 10,
        },
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
    {
        "NAME": "django_advanced_password_validation.advanced_password_validation.ContainsDigitsValidator",
        "OPTIONS": {"min_digits": 1},
    },
    {
        "NAME": "django_advanced_password_validation.advanced_password_validation.ContainsUppercaseValidator",
        "OPTIONS": {"min_uppercase": 1},
    },
    {
        "NAME": "django_advanced_password_validation.advanced_password_validation.ContainsLowercaseValidator",
        "OPTIONS": {"min_lowercase": 1},
    },
    {
        "NAME": "django_advanced_password_validation.advanced_password_validation.ContainsSpecialCharactersValidator",
        "OPTIONS": {"min_characters": 1},
    },
    {
        "NAME": "django_advanced_password_validation.advanced_password_validation.MaximumLengthValidator",
        "OPTIONS": {"max_length": 128},
    },
]

LANGUAGE_CODE = "en-us"
TIME_ZONE = "UTC"
USE_I18N = True
USE_L10N = True
USE_TZ = True
STATIC_URL = "/static/"
STATIC_ROOT = os.path.join(BASE_DIR, "staticfiles")


django_advanced_password_validation/tests/test_validators.py
"""
Tests for the advanced_password_validation module.
"""

import pytest
from django.contrib.auth.password_validation import validate_password
from django.core.exceptions import ValidationError

from ..advanced_password_validation import (
    ConsecutivelyDecreasingDigitValidator,
    ConsecutivelyIncreasingDigitValidator,
    ContainsDigitsValidator,
    ContainsLowercaseValidator,
    ContainsSpecialCharactersValidator,
    ContainsUppercaseValidator,
    MaxConsecutiveCharactersValidator,
    MaximumLengthValidator,
)


def test_contains_digits_validator():
    """
    Test that the ContainsDigitsValidator works as expected and raises a
    ValidationError when the password does not contain any digits.
    """
    validator = ContainsDigitsValidator()
    # None means that the password is valid. Otherwise it returns a ValidationError.
    assert validator.validate("1234567890") is None
    with pytest.raises(ValidationError) as exc:
        validator.validate("abcdefghij")
    assert exc.value.code == "password_too_weak"
    assert exc.value.message == "Password must contain at least 1 number."


def test_contains_digits_get_help_text():
    """
    Test that the get_help_text string works as expected.
    """
    validator = ContainsDigitsValidator(min_digits=1)
    assert validator.get_help_text() == "Your password must contain at least 1 number."
    validator = ContainsDigitsValidator(min_digits=2)
    assert validator.get_help_text() == "Your password must contain at least 2 numbers."


def test_contains_uppercase_validator():
    """
    Test that the ContainsUppercaseValidator works as expected and raises a
    ValidationError when the password does not contain any uppercase characters.
    """
    validator = ContainsUppercaseValidator()
    assert validator.validate("ABCDEFGHIJ") is None
    with pytest.raises(ValidationError) as exc:
        validator.validate("abcdefghij")
    assert exc.value.code == "password_too_weak"
    assert exc.value.message == "Password must contain at least 1 uppercase character."


def test_contains_uppercase_get_help_text():
    """
    Test that the get_help_text string works as expected.
    """
    validator = ContainsUppercaseValidator(min_uppercase=1)
    assert (
        validator.get_help_text()
        == "Your password must contain at least 1 uppercase character."
    )
    validator = ContainsUppercaseValidator(min_uppercase=2)
    assert (
        validator.get_help_text()
        == "Your password must contain at least 2 uppercase characters."
    )


def test_contains_lowercase_validator():
    """
    Test that the ContainsLowercaseValidator works as expected and raises a
    ValidationError when the password does not contain any lowercase characters.
    """
    validator = ContainsLowercaseValidator()
    assert validator.validate("abcdefghij") is None
    with pytest.raises(ValidationError) as exc:
        validator.validate("ABCDEFGHIJ")
    assert exc.value.code == "password_too_weak"
    assert exc.value.message == "Password must contain at least 1 lowercase character."


def test_contains_lowercase_get_help_text():
    """
    Test that the get_help_text string works as expected.
    """
    validator = ContainsLowercaseValidator(min_lowercase=1)
    assert (
        validator.get_help_text()
        == "Your password must contain at least 1 lowercase character."
    )
    validator = ContainsLowercaseValidator(min_lowercase=2)
    assert (
        validator.get_help_text()
        == "Your password must contain at least 2 lowercase characters."
    )


def test_contains_special_characters_validator():
    """
    Test that the ContainsSpecialCharactersValidator works as expected and raises a
    ValidationError when the password does not contain any special characters.
    """
    validator = ContainsSpecialCharactersValidator()
    assert validator.validate("!@#$%^&*()") is None
    with pytest.raises(ValidationError) as exc:
        validator.validate("abcdefghij")
    assert exc.value.code == "password_too_weak"
    assert (
        exc.value.message == "Password must contain at least 1 special character ("
        " !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~)."
    )


def test_contains_special_characters_get_help_text():
    """
    Test that the get_help_text string works as expected.
    """
    validator = ContainsSpecialCharactersValidator(min_characters=1)
    assert (
        validator.get_help_text()
        == "Your password must contain at least 1 special character ("
        " !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~)."
    )
    validator = ContainsSpecialCharactersValidator(min_characters=2)
    assert (
        validator.get_help_text()
        == "Your password must contain at least 2 special characters ("
        " !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~)."
    )


def test_maximum_length_validator():
    """
    Test that the MaximumLengthValidator works as expected and raises a
    ValidationError when the password is longer than the maximum length.
    """
    validator = MaximumLengthValidator(max_length=10)
    assert validator.validate("abcdefghij") is None
    with pytest.raises(ValidationError) as exc:
        validator.validate("abcdefghijk")
    assert exc.value.message == "Password must contain at maximum 10 characters."


def test_maximum_length_get_help_text():
    """
    Test that the get_help_text string works as expected.
    """
    validator = MaximumLengthValidator(max_length=12)
    assert (
        validator.get_help_text() == "Password must contain at maximum 12 characters."
    )


def test_max_consecutive_characters_validator():
    """
    Test that the MaxConsecutiveCharactersValidator works as expected and raises a
    ValidationError when the password contains more than the maximum number of
    consecutive characters.
    """
    validator = MaxConsecutiveCharactersValidator()
    assert validator.validate("abcdefghij") is None
    assert validator.validate("aaabbbccc") is None
    with pytest.raises(ValidationError) as exc:
        validator.validate("aaaabbbccc")
    assert (
        exc.value.message
        == "Password contains consecutively repeating characters. e.g 'aaa' or '111'"
    )


def test_max_consecutive_characters_get_help_text():
    """
    Test that the get_help_text string works as expected.
    """
    validator = MaxConsecutiveCharactersValidator()
    assert (
        validator.get_help_text()
        == "Password cannot contain consecutively repeating characters. e.g 'aaa' or"
        " '111'"
    )


def test_consecutively_increasing_digit_validator():
    """
    Test that the ConsecutivelyIncreasingDigitValidator works as expected and raises a
    ValidationError when the password contains more than the maximum number of
    consecutive characters.
    """
    validator = ConsecutivelyIncreasingDigitValidator(max_consecutive=3)
    assert validator.validate("abcdefghij") is None
    assert validator.validate("abcdefg123") is None
    with pytest.raises(ValidationError) as exc:
        validator.validate("1234567890")
    assert (
        exc.value.message
        == "Password contains consecutively increasing digits. e.g '12345'"
    )


def test_consecutively_increasing_digit_get_help_text():
    """
    Test that the get_help_text string works as expected.
    """
    validator = ConsecutivelyIncreasingDigitValidator(max_consecutive=3)
    assert (
        validator.get_help_text()
        == "Password cannot contain consecutively increasing digits. e.g '12345'"
    )


def test_consecutively_decreasing_digit_validator():
    """
    Test that the ConsecutivelyDecreasingDigitValidator works as expected and raises a
    ValidationError when the password contains more than the maximum number of
    consecutive characters.
    """
    validator = ConsecutivelyDecreasingDigitValidator(max_consecutive=3)
    assert validator.validate("abcdefghij") is None
    assert validator.validate("abcdefg321") is None
    with pytest.raises(ValidationError) as exc:
        validator.validate("9876543210")
    assert (
        exc.value.message
        == "Password contains consecutively decreasing digits. e.g '54321'"
    )


def test_consecutively_decreasing_digit_get_help_text():
    """
    Test that the get_help_text string works as expected.
    """
    validator = ConsecutivelyDecreasingDigitValidator(max_consecutive=3)
    assert (
        validator.get_help_text()
        == "Password cannot contain consecutively decreasing digits. e.g '54321'"
    )


def test_valid_password():
    """
    Test that the validate_password function works as expected.
    """
    assert validate_password("Abc$d1234!") is None


def test_invalid_password():
    """
    Test that the validate_password function works as expected and raises all
    relevant ValidationErrors when the password is too weak and doesn't conform
    to the password validation rules.
    """
    with pytest.raises(ValidationError) as exc:
        validate_password("")
    assert exc.value.messages == [
        "This password is too short. It must contain at least 10 characters.",
        "Password must contain at least 1 number.",
        "Password must contain at least 1 uppercase character.",
        "Password must contain at least 1 lowercase character.",
        (
            "Password must contain at least 1 special character ("
            " !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~)."
        ),
    ]


django_advanced_password_validation/advanced_password_validation.py
"""
Advanced password validation
"""

from django.core.exceptions import ValidationError
from django.utils.translation import ngettext_lazy
from django.utils.translation import gettext_lazy


class ContainsDigitsValidator:
    """
    Validates whether the password contains at least min_digits digits.
    """

    def __init__(self, min_digits=1):
        """Initializes the validator.

        Args:
            min_digits (int, optional): Minimum number of digits to validate password
                against. Defaults to 1.
        """
        self.min_digits = min_digits

    def validate(self, password, user=None):
        """
        Validates whether the password contains at least min_digits digits.

        Args:
            password (str): The password to validate.
            user (User): The user to validate the password for. (unused)

        Raises:
            ValidationError: Password must contain at least {self.min_digits} number(s).
        """
        if sum(c.isdigit() for c in password) < self.min_digits:
            raise ValidationError(
                ngettext_lazy(
                    "Password must contain at least %(min_digits)s number.",
                    "Password must contain at least %(min_digits)s numbers.",
                    self.min_digits,
                )
                % {"min_digits": self.min_digits},
                code="password_too_weak",
            )

    def get_help_text(self):
        """
        Get the help text for the validator.
        """
        return ngettext_lazy(
            "Your password must contain at least %(min_digits)s number.",
            "Your password must contain at least %(min_digits)s numbers.",
            self.min_digits,
        ) % {"min_digits": self.min_digits}


class ContainsUppercaseValidator:
    """
    Validates whether the password contains at least min_uppercase uppercase characters.
    """

    def __init__(self, min_uppercase=1):
        """Initializes the validator.

        Args:
            min_uppercase (int, optional): Minimum number of uppercase characters to
                validate password against. Defaults to 1.
        """
        self.min_uppercase = min_uppercase

    def validate(self, password, user=None):
        """
        Validates whether the password contains at least min_uppercase uppercase characters.

        Args:
            password (str): The password to validate.
            user (User): The user to validate the password for. (unused)

        Raises:
            ValidationError: Password must contain at least {self.min_uppercase} uppercase
                character(s).
        """
        if sum(c.isupper() for c in password) < self.min_uppercase:
            raise ValidationError(
                ngettext_lazy(
                    "Password must contain at least %(min_uppercase)s uppercase"
                    " character.",
                    "Password must contain at least %(min_uppercase)s uppercase"
                    " characters.",
                    self.min_uppercase,
                )
                % {"min_uppercase": self.min_uppercase},
                code="password_too_weak",
            )

    def get_help_text(self):
        """
        Get the help text for the validator.
        """
        return ngettext_lazy(
            "Your password must contain at least %(min_uppercase)s uppercase"
            " character.",
            "Your password must contain at least %(min_uppercase)s uppercase"
            " characters.",
            self.min_uppercase,
        ) % {"min_uppercase": self.min_uppercase}


class ContainsLowercaseValidator:
    """
    Validates whether the password contains at least min_lowercase lowercase characters.
    """

    def __init__(self, min_lowercase=1):
        """Initializes the validator.

        Args:
            min_lowercase (int, optional): Minimum number of lowercase characters to
                validate password against. Defaults to 1.
        """
        self.min_lowercase = min_lowercase

    def validate(self, password, user=None):
        """
        Validates whether the password contains at least min_lowercase lowercase characters.

        Args:
            password (str): The password to validate.
            user (User): The user to validate the password for. (unused)

        Raises:
            ValidationError: Password must contain at least {self.min_lowercase} lowercase
                character(s).
        """
        if sum(c.islower() for c in password) < self.min_lowercase:
            raise ValidationError(
                ngettext_lazy(
                    "Password must contain at least %(min_lowercase)s lowercase"
                    " character.",
                    "Password must contain at least %(min_lowercase)s lowercase"
                    " characters.",
                    self.min_lowercase,
                )
                % {"min_lowercase": self.min_lowercase},
                code="password_too_weak",
            )

    def get_help_text(self):
        """
        Get the help text for the validator.
        """
        return ngettext_lazy(
            "Your password must contain at least %(min_lowercase)s lowercase"
            " character.",
            "Your password must contain at least %(min_lowercase)s lowercase"
            " characters.",
            self.min_lowercase,
        ) % {"min_lowercase": self.min_lowercase}


class ContainsSpecialCharactersValidator:
    """
    Validates whether the password contains at least min_characters special characters.
    """

    def __init__(self, min_characters=1):
        """Initializes the validator.

        Args:
            min_characters (int, optional): Minimum number of special characters to
                validate password against. Defaults to 1.
        """
        self.min_characters = min_characters
        self.characters = " !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"

    def validate(self, password, user=None):
        """
        Validates whether the password contains at least min_characters special characters.

        Args:
            password (str): The password to validate.
            user (User): The user to validate the password for. (unused)

        Raises:
            ValidationError: Password must contain at least {self.min_characters} special
                character(s).
        """
        if sum(c in set(self.characters) for c in password) < self.min_characters:
            raise ValidationError(
                ngettext_lazy(
                    "Password must contain at least %(min_characters)s special"
                    " character (%(special_characters)s).",
                    "Password must contain at least %(min_characters)s special"
                    " characters (%(special_characters)s).",
                    self.min_characters,
                )
                % {
                    "min_characters": self.min_characters,
                    "special_characters": "".join(self.characters),
                },
                code="password_too_weak",
            )

    def get_help_text(self):
        """
        Get the help text for the validator.
        """
        return ngettext_lazy(
            "Your password must contain at least %(min_characters)s special character"
            " (%(special_characters)s).",
            "Your password must contain at least %(min_characters)s special characters"
            " (%(special_characters)s).",
            self.min_characters,
        ) % {
            "min_characters": self.min_characters,
            "special_characters": "".join(self.characters),
        }


class MaximumLengthValidator:
    """
    OWASP recommends setting a maximum password length, typically 128 characters, to prevent
    'long password Denial of Service attacks'.
    """

    def __init__(self, max_length=128):
        """Initializes the validator.

        Args:
            max_length (int, optional): Maximum length of the password. Defaults to 128.
        """
        self.max_length = max_length

    def validate(self, password, user=None):
        """
        Validates whether the password contains at least min_characters special characters.

        Args:
            password (str): The password to validate.
            user (User): The user to validate the password for. (unused)

        Raises:
            ValidationError: Password must contain at least {self.min_characters} special
                character(s).
        """
        if len(password) > self.max_length:
            raise ValidationError(
                ngettext_lazy(
                    "Password must contain at maximum %(max_length)s character.",
                    "Password must contain at maximum %(max_length)s characters.",
                    self.max_length,
                )
                % {"max_length": self.max_length}
            )

    def get_help_text(self):
        """
        Get the help text for the validator.
        """
        return ngettext_lazy(
            "Password must contain at maximum %(max_length)s character.",
            "Password must contain at maximum %(max_length)s characters.",
            self.max_length,
        ) % {"max_length": self.max_length}


class MaxConsecutiveCharactersValidator:
    """
    Validates whether the password contains more than max_consecutive consecutive characters.
    """

    def __init__(self, max_consecutive=3):
        """Initializes the validator.

        Args:
            max_consecutive (int, optional): Maximum number of consecutive characters to
                validate password against. Defaults to 3.
        """
        self.max_consecutive = max_consecutive

    def validate(self, password, user=None):
        """
        Validates whether the password contains more than max_consecutive consecutive characters.

        Args:
            password (str): The password to validate.
            user (User): The user to validate the password for. (unused)

        Raises:
            ValidationError: Password cannot contain consecutively repeating
                characters. e.g 'aaa' or '111'
        """
        for c in password:
            if password.count(c) >= self.max_consecutive:
                check = c * (self.max_consecutive + 1)
                if check in password:
                    raise ValidationError(
                        gettext_lazy(
                            "Password contains consecutively repeating characters. "
                            "e.g 'aaa' or '111'"
                        )
                    )

    def get_help_text(self):
        """
        Get the help text for the validator.
        """
        return gettext_lazy(
            "Password cannot contain consecutively repeating characters. e.g 'aaa' or"
            " '111'"
        )


class ConsecutivelyIncreasingDigitValidator:
    """
    Validates whether the password contains consecutively increasing digits.
    """

    def __init__(self, max_consecutive=3):
        """Initializes the validator.

        Args:
            max_consecutive (int, optional): Maximum number of consecutive digits to
                validate password against. Defaults to 3.
        """
        self.max_consecutive = max_consecutive

    def validate(self, password, user=None):
        """
        Validates whether the password contains consecutively increasing digits.

        Args:
            password (str): The password to validate.
            user (User): The user to validate the password for. (unused)

        Raises:
            ValidationError: Password contains consecutively increasing digits. e.g '12345'
        """
        for c in password:
            if c.isdigit():
                count = 0
                digit = int(c)
                index = password.index(c)

                try:
                    for i in range(1, self.max_consecutive + 1):
                        if password[index + i].isdigit():
                            if int(password[index + i]) == digit + 1:
                                count += 1
                                digit += 1

                                while count >= self.max_consecutive:
                                    raise ValidationError(
                                        gettext_lazy(
                                            "Password contains consecutively increasing"
                                            " digits. e.g '12345'"
                                        )
                                    )
                except IndexError:
                    pass

    def get_help_text(self):
        """
        Get the help text for the validator.
        """
        return gettext_lazy(
            "Password cannot contain consecutively increasing digits. e.g '12345'"
        )


class ConsecutivelyDecreasingDigitValidator:
    """
    Validates whether the password contains consecutively decreasing digits.
    """

    def __init__(self, max_consecutive=3):
        """Initializes the validator.

        Args:
            max_consecutive (int, optional): Maximum number of consecutive digits to
                validate password against. Defaults to 3.
        """
        self.max_consecutive = max_consecutive

    def validate(self, password, user=None):
        """
        Validates whether the password contains consecutively decreasing digits.

        Args:
            password (str): The password to validate.
            user (User): The user to validate the password for. (unused)

        Raises:
            ValidationError: Password contains consecutively decreasing digits. e.g '54321'
        """
        for c in password:
            if c.isdigit():
                count = 0
                digit = int(c)
                index = password.index(c)

                try:
                    for i in range(1, self.max_consecutive + 1):
                        if password[index + i].isdigit():
                            if int(password[index + i]) == digit - 1:
                                count += 1
                                digit -= 1

                                while count >= self.max_consecutive:
                                    raise ValidationError(
                                        gettext_lazy(
                                            "Password contains consecutively decreasing"
                                            " digits. e.g '54321'"
                                        )
                                    )
                except IndexError:
                    pass

    def get_help_text(self):
        """
        Get the help text for the validator.
        """
        return gettext_lazy(
            "Password cannot contain consecutively decreasing digits. e.g '54321'"
        )
